src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -266,8 +269,10 @@ private bool MatchRouteValue(ActionDescriptor action, MvcEndpointInfo endpointIn
             {
                 // Action does not have a value for this routeKey, most likely because action is not in an area
                 // Check that the pattern does not have a parameter for the routeKey
-                var matchingParameter = endpointInfo.ParsedPattern.Parameters.SingleOrDefault(p => string.Equals(p.Name, routeKey, StringComparison.OrdinalIgnoreCase));
-                if (matchingParameter == null)
+                var matchingParameter = endpointInfo.ParsedPattern.GetParameter(routeKey);
+                if (matchingParameter == null &&
+                    (!endpointInfo.ParsedPattern.Defaults.TryGetValue(routeKey, out var value) ||
+                    !string.IsNullOrEmpty(Convert.ToString(value))))
```

Comment:
This code wasn't checking the defaults for some reason. Basically if a require route value is null, then it should **not** be a value that can be supplied by the route.